src/server/gc_worker/gc_worker.rs

```diff
@@ -167,238 +155,108 @@ impl<E: Engine> GcRunner<E> {
-                .map_err(Error::from)
-                .map(|(keys, next)| {
-                    if keys.is_empty() {
-                        assert!(next.is_none());
-                        if is_range_start {
-                            GC_EMPTY_RANGE_COUNTER.inc();
-                        }
-                    }
-                    (keys, next)
-                })
-        };
-        self.stats.add(reader.get_statistics());
-        res
-    }
-
     /// Cleans up outdated data.
-    fn gc_keys(
-        &mut self,
-        ctx: &mut Context,
-        safe_point: TimeStamp,
-        keys: Vec<Key>,
-        mut next_scan_key: Option<Key>,
-    ) -> Result<Option<Key>> {
-        let snapshot = self.get_snapshot(ctx)?;
-        let mut txn = MvccTxn::for_scan(
-            snapshot,
-            Some(ScanMode::Forward),
-            TimeStamp::zero(),
-            !ctx.get_not_fill_cache(),
-        );
-        for k in keys {
-            let gc_info = txn.gc(k.clone(), safe_point)?;
-
-            if gc_info.found_versions >= GC_LOG_FOUND_VERSION_THRESHOLD {
-                debug!(
-                    "GC found plenty versions for a key";
-                    "region_id" => ctx.get_region_id(),
-                    "versions" => gc_info.found_versions,
-                    "key" => %k
-                );
-            }
-            // TODO: we may delete only part of the versions in a batch, which may not beyond
-            // the logging threshold `GC_LOG_DELETED_VERSION_THRESHOLD`.
-            if gc_info.deleted_versions as usize >= GC_LOG_DELETED_VERSION_THRESHOLD {
-                debug!(
-                    "GC deleted plenty versions for a key";
-                    "region_id" => ctx.get_region_id(),
-                    "versions" => gc_info.deleted_versions,
-                    "key" => %k
-                );
-            }
+    fn gc_key(&mut self, safe_point: TimeStamp, key: &Key) -> Result<()> {
+        let mut gc_info = GcInfo::default();
+        while !gc_info.is_completed {
+            let mut txn = MvccTxn::for_scan(
+                self.engine.snapshot_on_kv_engine(b"", b"")?,
+                Some(ScanMode::Forward),
+                TimeStamp::zero(),
+                false,
+            );
 
-            if !gc_info.is_completed {
-                next_scan_key = Some(k);
-                break;
+            let next_gc_info = txn.gc(key.clone(), safe_point).unwrap();
+            gc_info.found_versions += next_gc_info.found_versions;
+            gc_info.deleted_versions += next_gc_info.deleted_versions;
+            gc_info.is_completed = next_gc_info.is_completed;
+            self.stats.add(&txn.take_statistics());
+
+            let write_size = txn.write_size();
+            let modifies = txn.into_modifies();
+            if !modifies.is_empty() {
+                self.limiter.blocking_consume(write_size);
+                self.engine.modify_on_kv_engine(modifies)?;
```

Comment:
I think for most cases we won't delete too many versions for a single key. Previously we can batch multiple keys, but it seems now it won't. I'm afraid GC speed might be significantly reduced. How do you think?